### PR TITLE
Feat rpaasv2 logs

### PR DESCRIPTION
**Pull request by arthurcgc**

### CODE REVIEW

File: internal/pkg/rpaas/k8s.go [hunk 1]
```diff
@@ -125,6 +128,64 @@ func (q *fixedSizeQueue) Next() *remotecommand.TerminalSize {
...
+
+	podsInfo, err := m.getPods(ctx, nginx)
+	if err != nil {
+		return err
+	}
+
+	var candidate v1.Pod
+	for _, pod := range podsInfo {
+		if pod.Name == args.Pod && pod.Status.Phase == "Running" {
+			candidate = pod
```
Comment (wpjunior) [inline internal/pkg/rpaas/k8s.go:150]: @arthurcgc could you notify the user that have selected that pod ?, for example, some users may follow the behavior of tsuru apps, that receive all pods into a one command.

Example:
Write("Selected pod %s, if you want to check logs of other pods you should select the name of pod")
time.Sleep(time.Secod)

the code above could placed on client-side. If you want to inform the name of pod, you could use a header to carry, cause the body is a buffer.

Comment (arthurcgc) [reply internal/pkg/rpaas/k8s.go]: Yup, you're right... I'll add it to the command description and before the command execution. You're also right about the request body, it's already changed 😄

File: cmd/plugin/rpaasv2/cmd/log.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
+package cmd
```
Comment (nettoclaudio) [inline cmd/plugin/rpaasv2/cmd/log.go:1]: Missing license header.

```suggestion
// Copyright 2021 tsuru authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package cmd
```

File: cmd/plugin/rpaasv2/cmd/log_test.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
+package cmd
```
Comment (nettoclaudio) [inline cmd/plugin/rpaasv2/cmd/log_test.go:1]:
```suggestion
// Copyright 2021 tsuru authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package cmd
```

File: internal/pkg/rpaas/k8s.go [hunk 4]
```diff
@@ -125,6 +128,92 @@ func (q *fixedSizeQueue) Next() *remotecommand.TerminalSize {
...
+	}
+	if len(podsInfo) == 0 {
+		return ValidationError{Msg: "nginx instance has no pods associated to it"}
+	}
+
+	var candidate v1.Pod
+	found := false
+	for _, pod := range podsInfo {
+		if args.Pod != "" {
+			switch args.Container == "" {
```
Comment (nettoclaudio) [inline internal/pkg/rpaas/k8s.go:163]: I'd set the Container name as `nginx` if not provided.

```golang
containerName := args.Container
if containerName == "" {
    containerName = "nginx"
}
```

File: internal/pkg/rpaas/k8s.go [hunk 5]
```diff
@@ -125,6 +128,92 @@ func (q *fixedSizeQueue) Next() *remotecommand.TerminalSize {
...
+			case true:
+				if args.Pod == pod.Name && pod.Status.Phase == "Running" {
```
Comment (nettoclaudio) [inline internal/pkg/rpaas/k8s.go:165]: Why do we need ensure the pod phase as running to see its logs?

File: internal/web/log.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
+package web
```
Comment (nettoclaudio) [inline internal/web/log.go:1]: Missing license header.

```suggestion
// Copyright 2021 tsuru authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package web
```

File: internal/web/log_test.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
+package web
```
Comment (nettoclaudio) [inline internal/web/log_test.go:1]:
```suggestion
// Copyright 2021 tsuru authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package web
```

File: pkg/rpaas/client/log.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+func (c *client) Log(ctx context.Context, args LogArgs) error {
+	values := url.Values{
+		"follow": []string{strconv.FormatBool(args.Follow)},
+	}
+	pathName := fmt.Sprintf("/resources/%s/log?%s", args.Instance, values.Encode())
+	req, err := c.newRequest("GET", pathName, nil, args.Instance)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
```
Comment (nettoclaudio) [inline pkg/rpaas/client/log.go:22]: You should remove this as the employed HTTP method is GET.

File: pkg/rpaas/client/log.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
+package client
```
Comment (nettoclaudio) [inline pkg/rpaas/client/log.go:1]:
```suggestion
// Copyright 2021 tsuru authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package client
```

**Review by nettoclaudio — CHANGES_REQUESTED** [review]
I left tiny change requests. PTAL

BTW very nice work.

File: cmd/plugin/rpaasv2/cmd/log.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+			},
+			&cli.IntFlag{
+				Name:     "lines",
+				Aliases:  []string{"l"},
+				Usage:    "number of earlier log lines to show",
+				Required: false,
+			},
+			&cli.IntFlag{
+				Name:     "since",
+				Usage:    "only return logs newer than a relative duration like 5s, 2m, or 3h. Defaults to all logs.",
```
Comment (nettoclaudio) [inline cmd/plugin/rpaasv2/cmd/log.go:49]:
```suggestion
				Usage:    "only return logs newer than a relative duration like 5s, 2m, or 3h. Defaults to last 24 hours.",
```

File: cmd/plugin/rpaasv2/cmd/log.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+import (
+	rpaasclient "github.com/tsuru/rpaas-operator/pkg/rpaas/client"
+	"github.com/urfave/cli/v2"
+)
+
+func NewCmdLogs() *cli.Command {
+	return &cli.Command{
+		Name:    "logs",
+		Usage:   "Fetches logs from a rpaasv2 instance pod",
```
Comment (nettoclaudio) [inline cmd/plugin/rpaasv2/cmd/log.go:15]:
```suggestion
		Usage:   "Fetches and prints logs from instance pods",
```

File: cmd/plugin/rpaasv2/cmd/log.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+				Usage:    "specific container to log from",
+				Required: false,
+			},
+			&cli.IntFlag{
+				Name:     "lines",
+				Aliases:  []string{"l"},
+				Usage:    "number of earlier log lines to show",
+				Required: false,
+			},
+			&cli.IntFlag{
```
Comment (nettoclaudio) [inline cmd/plugin/rpaasv2/cmd/log.go:47]: Duration fits better.

```suggestion
			&cli.DurationFlag{
```

File: internal/pkg/rpaas/k8s.go [hunk 13]
```diff
@@ -126,6 +129,47 @@ func (q *fixedSizeQueue) Next() *remotecommand.TerminalSize {
...
+func defaultSternTemplate() (*template.Template, error) {
+	t := "{{color .PodColor .PodName}} {{color .ContainerColor .ContainerName}} {{.Message}}\r\n"
+	funcs := map[string]interface{}{
+		"json": func(in interface{}) (string, error) {
+			b, err := json.Marshal(in)
+			if err != nil {
+				return "", err
+			}
+			return string(b), nil
+		},
```
Comment (nettoclaudio) [inline internal/pkg/rpaas/k8s.go:141]: Unused code.
```suggestion
```

File: internal/pkg/rpaas/log.go [hunk 14]
```diff
@@ -0,0 +1,167 @@
+package rpaas
```
Comment (nettoclaudio) [inline internal/pkg/rpaas/log.go:1]: Missing license header.

File: internal/pkg/rpaas/k8s.go [hunk 15]
```diff
@@ -126,6 +129,47 @@ func (q *fixedSizeQueue) Next() *remotecommand.TerminalSize {
 	return q.sz
 }
 
+func defaultSternTemplate() (*template.Template, error) {
+	t := "{{color .PodColor .PodName}} {{color .ContainerColor .ContainerName}} {{.Message}}\r\n"
```
Comment (nettoclaudio) [inline internal/pkg/rpaas/k8s.go:133]: We should consider to add a flag to disable the coloring.

File: internal/web/log.go [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+	params := c.Request().URL.Query()
+	var err error
+
+	var pLines *int64
+	lines, _ := strconv.ParseInt(params.Get("lines"), 10, 64)
+	if lines > 0 {
+		pLines = &lines
+	}
+
+	tSince := int64(172800) // last 48 hours by default -> 48 * 60(minutes) * 60(seconds)
```
Comment (nettoclaudio) [inline internal/web/log.go:83]: Last 48h might return too much logs, let's try 24h.

```suggestion
	tSince := int64(24 * 60 * 60) // last 24 hours by default
```

**Review by nettoclaudio — APPROVED** [review]

**Pull request merged**